apps/dot_ext/forms.py

```diff
@@ -86,58 +107,201 @@ def clean(self):
+        )
+
+    def clean_name(self):
+        print("clean name")
+
+        name = self.cleaned_data.get("name")
+        print(str(name))
+        app_model = get_application_model()
+        if (
+            app_model.objects.filter(name__iexact=name)
+            .exclude(pk=self.instance.pk)
+            .exists()
+        ):
+            raise forms.ValidationError(
+                """
+                                        It looks like this application name
+                                        is already in use with another app.
+                                        Please enter a different application
+                                        name to prevent future errors.
+                                        Note that names are case-insensitive.
+                                        """
+            )
+        return name
+
+    def clean_logo_image(self):
+        print("clean logo image")
+
+        logo_image = self.cleaned_data.get("logo_image")
+        print(str(logo_image))
+        if getattr(logo_image, "name", False):
+            validate_logo_image(logo_image)
+        return logo_image
+
+    def clean_redirect_uris(self):
+        print("clean redirect uris")
+
+        redirect_uris = self.cleaned_data.get("redirect_uris")
+        print(str(redirect_uris))
+        if getattr(settings, "BLOCK_HTTP_REDIRECT_URIS", True):
+            if redirect_uris:
+                for u in redirect_uris.split():
+                    if u.startswith("http://"):
+                        msg = _("Redirect URIs must not use http.")
+                        raise forms.ValidationError(msg)
+        return redirect_uris
+
+    def clean_require_demographic_scopes(self):
+        print("clean demo scopes")
+
+        require_demographic_scopes = self.cleaned_data.get("require_demographic_scopes")
+        print(str(require_demographic_scopes))
+        if type(require_demographic_scopes) != bool:
+            msg = _(
+                "Does your application need to collect beneficary demographic information must be (Yes/No)."
+            )
+            raise forms.ValidationError(msg)
+        return require_demographic_scopes
+
+    def save(self, *args, **kwargs):
+        app = self.instance
+        print("save")
+        print(str(self.cleaned_data))
+
+        new_user_model = User.objects.create(
+            username=self.cleaned_data.get("name") + "@example.com",
+            password=str(uuid.uuid4()),
+            is_active=True,
+        )
+        group = Group.objects.get(name="BlueButton")
+        new_user_model.groups.add(group)
+        new_user_model.save()
+
+        UserProfile.objects.create(
+            user=new_user_model,
+            organization_name=self.cleaned_data.get("organization_name"),
+        )
+
+        app = super().save(*args, **kwargs)
+        app.agree = True
+        app.user = new_user_model
+        app.authorization_grant_type = Application.GRANT_AUTHORIZATION_CODE
```

Comment:
Setting the `app.client_type = Application.CLIENT_CONFIDENTIAL` can be done here.

Reply:
It would be good to setup the `scope` here, so that it doesn't need selected manually later.

This would be similar to the following:

`app.scope.add("profile", "patient/Patient.read", "patient/Coverage.read", "patient/ExplanationOfBenefit.read")`

Could add a variable to settings/base.py with this maybe:

APPLICATION_DEFAULT_SCOPES = ["profile", "patient/Patient.read", "patient/Coverage.read", "patient/ExplanationOfBenefit.read"]

Then set `app.scope = APPLICATION_DEFAULT_SCOPES`

Reply:
Currently, the application label gets setup via a separate admin page.  A chooser for the application labels for the new application might be good to add here to save a step.  That's for the `ApplicationLabel` model labels.

Reply:
@dtisza1 I thought answering this question: `Are demographic scopes required?` would basically automatically handle the scopes? At least from the perspective we really care about?

Reply:
Yes to all this above.  Anything we can do to make the process of credentialing easier.

Reply:
@nbragdon You were right about the `app.scopes` not actually getting used in favor of a single scope via the `application.require_demographic_scopes` flag.

There is just one minor issue with having no `application.scope` set.  When editing an app via the MyApplication admin view, it will produce an error when saving---unless at least one scope is selected.

I think a backlog story to refactor & simplify app scopes would be good to do in the future.

Reply:
This change looks good!